### PR TITLE
[AvroTensorDataset] Add AvroTensorDataset to allow data conversion from avro to Tensorflow tensors

### PR DESCRIPTION
**Pull request by lijuanzhang78**
### Context
LinkedIn is open sourcing AvroTensorDataset, a TensorFlow dataset for reading, parsing, and processing Avro data. At LinkedIn, TensorFlow is the de facto framework for machine learning training. Workloads ingest terabytes of training data residing on HDFS into memory. Ideally, training workloads are compute-bound (since compute hardware is expensive), but historically we found some of our workloads were I/O bound. Hence, we built AvroTensorDataset to address the performance gap. 

### Code structure
Core logic is added to core/kernels/avro/atds, which includes the following:

- AvroBlockReader

- Decoders (ATDS decoder, dense feature decoder, sparse feature decoder, varlen feature decoder)

- Compression handler

- Shuffle handler

- ATDSDataset kernel

ATDS Python API is added to python/experiment/atds

A bunch of tests are added to tests/atds_avro and tests/benchmark

### Performance
We benchmarked AvroTensorDataset against the existing AvroRecordDataset under various batch sizes with the following parameters:
Number of records: 303062
Number of files: 10
Number of feature columns: 15 / 19 (used / total)

<meta charset="utf-8"><b style="font-weight:normal;" id="docs-internal-guid-866e5792-7fff-68b0-1fa5-cfe0d91e4f89"><p dir="ltr" style="line-height:1.38;margin-top:0pt;margin-bottom:0pt;"><span style="font-size:11pt;font-family:Arial;color:#000000;background-color:transparent;font-weight:400;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;white-space:pre-wrap;">Results:</span></p><div dir="ltr" style="margin-left:0pt;" align="left">

  | Batch size 64 | Batch size 256 | Batch size 1024
-- | -- | -- | --
AvroRecordDataset (baseline) | 40 ms / step | 160 ms / step | 650 ms / step
AvroTensorDataset | 1.2 ms / step | 1.3 ms / step | 4 ms / step
Improvement | 33x | 123x | 162x

</div></b>

### Acknowledgement
AvroTensorDataset is based on Linkedin's internal implementation pioneered by @hungj (Jonathan Hung) , @pliao (Pei-lun Liao), @ashahab (Abin Shahab)

### Documentation 
[Next Gen Avro Data Pipeline for Deep Learning](https://docs.google.com/presentation/d/1CTaiOlMHvF1yhHroCxBsGwGfpEELngVeENisJphcUZ0/edit#slide=id.g125ec881092_0_0)

### CODE REVIEW

**Comment by plliao** [comment]
Title is a little bit misleading. It is not TFRecord but TensorFlow tensors.

**Comment by kvignesh1420** [comment]
@lijuanzhang78 the CLA check is failing. Can you force-push to the branch after signing the CLA?

**Comment by kvignesh1420** [comment]
Also, is it possible to break this PR into multiple PRs? Reviewing ~15K lines in a single PR is not ideal IMHO. Thank you for your contribution!

**Comment by lijuanzhang78** [comment]
@yongtang Hi! I saw you manually triggered the CI workflow. I am wondering if PR author is allowed to trigger it somehow? Another question is when we can expect a docker image update with Tensorflow 2.12? Thanks in advance!

**Comment by lijuanzhang78** [comment]
> @lijuanzhang78 the CLA check is failing. Can you force-push to the branch after signing the CLA?

@kvignesh1420 Hi! CLA has passed. Can you help approve the workflow? Thank you!

**Comment by kvignesh1420** [comment]
@lijuanzhang78 I have approved the flow for now to check the status of CI jobs. Also, as discussed today, it would be great if you can modify this PR with the functionality code + basic tests (A new PR is also fine).

**Comment by lijuanzhang78** [comment]
> @lijuanzhang78 I have approved the flow for now to check the status of CI jobs. Also, as discussed today, it would be great if you can modify this PR with the functionality code + basic tests (A new PR is also fine). Thanks again!

Thanks for the PR sync-up! Per the discussion, the following basic tests are added for the first PR:
```
tests/test_atds_avro/benchmark/test_dense_tensor_atds_benchmark.py
tests/test_atds_avro/benchmark/test_sparse_tensor_atds_benchmark.py
tests/test_atds_avro/benchmark/test_varlen_tensor_atds_benchmark.py
```

**Comment by kvignesh1420** [comment]
Thanks @lijuanzhang78, will take a look.

File: tensorflow_io/python/experimental/benchmark/data_source.py
Comment (kvignesh1420) [inline tensorflow_io/python/experimental/benchmark/data_source.py:1]: The content of this file seems to be a duplicate of `benchmark/data_source.py`. Can we reuse the code and remove the duplicate classes?

**Review by kvignesh1420 — COMMENTED** [review]
Hi,

Here are some minor code organization-based comments to get started with the review.

1. I think `python/experimental` is not the right place for the `benchmark` functionality. It seems like the purpose of this sub-package is to generate test files. If that is the case, then `tests/test_atds_avro` seems to be a suitable place for it. The same applies to `tests/test_benchmark`.
2. Please check [io/docs/development.md](https://github.com/tensorflow/io/blob/master/docs/development.md) to fix all lint errors.
3. Please check individual comments in respective files.

File: tensorflow_io/core/kernels/avro/atds/decoder_base.h [hunk 1]
```diff
@@ -0,0 +1,62 @@
+#ifndef TENSORFLOW_DATA_CORE_KERNELS_AVRO_ATDS_DECODER_BASE_H_
+#define TENSORFLOW_DATA_CORE_KERNELS_AVRO_ATDS_DECODER_BASE_H_
```
[the remark this answers is not in the document]
Comment (lijuanzhang78) [reply tensorflow_io/core/kernels/avro/atds/decoder_base.h]: thanks for the catch!

File: tensorflow_io/core/kernels/avro/atds/decoder_base.h [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+template<typename Metadata>
+std::unique_ptr<DecoderBase> CreateFeatureDecoder(const avro::NodePtr&, const Metadata&);
+
+template<typename Metadata>
+Status ValidateSchema(const avro::NodePtr&, const Metadata&);
+
+}  // namespace atds
+}  // namespace tensorflow
+
+#endif // TENSORFLOW_DATA_CORE_KERNELS_AVRO_ATDS_DECODER_BASE_H_
```
[the remark this answers is not in the document]
Comment (lijuanzhang78) [reply tensorflow_io/core/kernels/avro/atds/decoder_base.h]: Updated.

File: tensorflow_io/python/experimental/benchmark/data_source.py
Comment (lijuanzhang78) [reply tensorflow_io/python/experimental/benchmark/data_source.py:1]: > Hi,
> 
> Here are some minor code organization-based comments to get started with the review.
> 
> 1. I think `python/experimental` is not the right place for the `benchmark` functionality. It seems like the purpose of this sub-package is to generate test files. If that is the case, then `tests/test_atds_avro` seems to be a suitable place for it. The same applies to `tests/test_benchmark`.
> 2. Please check [io/docs/development.md](https://github.com/tensorflow/io/blob/master/docs/development.md) to fix all lint errors.
> 3. Please check individual comments in respective files.

Good point! The test utils are moved from `tensorflow_io/python/experimental/benchmark` to `tests/test_atds_avro/utils`. And the folder tests/test_benchmark is for benchmark with TFRecord in our code base, I am fine not to add them in the first PR. WDYT? @plliao @hungj

File: tensorflow_io/core/kernels/avro/atds/shuffle_handler.h [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+namespace tensorflow {
+namespace data {
+
+class ShuffleHandler {
+ public:
+  ShuffleHandler(mutex* mu) {
+    mu_ = mu;
+    ResetRngs();
+  }
+  void SampleBlocks(size_t batch_size, bool shuffle,
```
Comment (kvignesh1420) [inline tensorflow_io/core/kernels/avro/atds/shuffle_handler.h:35]: I went through the shuffle algorithm which was posted in the FDSDataset Google doc. The purpose of sampling records randomly from different blocks is mentioned as
```
 In this way, the data from the latter blocks can be collected 
earlier.
```
I am assuming, you mean that the shuffle op can be completed within the prefetch phase itself? (please correct me if I misunderstood) What was the performance gap when using the standard `tf.data` prefetching and shuffle ops and disabling this custom shuffling?

File: tensorflow_io/python/experimental/atds/dataset.py [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+            elif isinstance(feature, SparseFeature):
+                feature_types.append(_SPARSE_FEATURE_TYPE)
+                sparse_dtypes.append(feature.dtype)
+                sparse_shapes.append(shape)
+                element_spec[key] = tf.SparseTensorSpec(shape, feature.dtype)
+            elif isinstance(feature, VarlenFeature):
+                feature_types.append(_VARLEN_FEATURE_TYPE)
+                sparse_dtypes.append(feature.dtype)
+                sparse_shapes.append(shape)
+                element_spec[key] = tf.SparseTensorSpec(shape, feature.dtype)
```
[the remark this answers is not in the document]
Comment (lijuanzhang78) [reply tensorflow_io/python/experimental/atds/dataset.py:255]: Discussed with team, we don't support RaggedTensor internally. Ragged tensor is a super set of varlen. For example, varlen can only have 1 unknown dimension but ragged can be n-dimensiond. However, the structure is flexible for others to contribute ragged tensor if they need. cc: @plliao

File: tests/test_atds_avro/benchmark/test_dense_tensor_atds_benchmark.py [hunk 5]
```diff
@@ -0,0 +1,313 @@
+# Copyright 2021 The TensorFlow Authors. All Rights Reserved.
```
[the remark this answers is not in the document]
Comment (lijuanzhang78) [reply tests/test_atds_avro/benchmark/test_dense_tensor_atds_benchmark.py]: updated

File: tensorflow_io/python/experimental/atds/atds_writer.py [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+    VarLenTensorGeneratorBase,
+)
+from tensorflow_io.python.experimental.atds.features import (
+    DenseFeature,
+    SparseFeature,
+    VarlenFeature,
+)
+
+
+class ATDSWriter(FileWriter):
```
[the remark this answers is not in the document]
Comment (lijuanzhang78) [reply tensorflow_io/python/experimental/atds/atds_writer.py]: Makes sense. It is moved to tests directory.

File: tensorflow_io/python/experimental/atds/dataset.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (kvignesh1420) [reply tensorflow_io/python/experimental/atds/dataset.py:269]: ok

File: tensorflow_io/core/kernels/avro/atds/decoder_test_util.h [hunk 7]
```diff
@@ -0,0 +1,404 @@
...
+  auto& feature_values = feature.value();
+  for (byte_array value : values) {
+    feature_values.emplace_back(value);
+  }
+}
+
+template <typename T>
+void AddSparseValue(avro::GenericDatum& datum, const string& name,
+                    const std::vector<std::vector<long>>& indices,
+                    const std::vector<T>& values) {
```
Comment (kvignesh1420) [inline tensorflow_io/core/kernels/avro/atds/decoder_test_util.h:149]: @lijuanzhang78 there is a build error on mac that is caused due to this template function:

```bash
2023-04-22T18:39:56.1638890Z tensorflow_io/core/kernels/avro/atds/atds_decoder_test.cc:76:3: error: no matching function for call to 'AddSparseValue'
2023-04-22T18:39:56.1639280Z   AddSparseValue(atds_datum, feature_names[3], sparse_int_1d_indices,
2023-04-22T18:39:56.1639530Z   ^~~~~~~~~~~~~~
2023-04-22T18:39:56.1640170Z ./tensorflow_io/core/kernels/avro/atds/decoder_test_util.h:147:6: note: candidate function template not viable: no known conversion from 'vector<vector<int64_t>>' to 'const vector<vector<long>>' for 3rd argument
2023-04-22T18:39:56.1640640Z void AddSparseValue(avro::GenericDatum& datum, const string& name,
2023-04-22T18:39:56.1640880Z      ^
```

Can you modify the test case to handle this issue?

Comment (lijuanzhang78) [reply tensorflow_io/core/kernels/avro/atds/decoder_test_util.h:149]: @kvignesh1420 Hi! The mac build should be good now.

**Comment by kvignesh1420** [comment]
@lijuanzhang78 thanks for the changes. Overall the functionality looks good to me and tests are failing due to an unrelated GCS error. Can you please fix the lint error before I merge this PR?

**Comment by lijuanzhang78** [comment]
> @lijuanzhang78 thanks for the changes. Overall the functionality looks good to me and tests are failing due to an unrelated GCS error. Can you please fix the lint error before I merge this PR?

cool, thanks! lint error fixed. but I saw build on windows is failing. Is it a known issue?

**Comment by kvignesh1420** [comment]
> 

Yes, it is an ongoing issue. There are unresolved symbols for other ops as well.
```
2023-04-25T04:32:29.7821194Z bigquery_ops.lo.lib(bigquery_dataset_op.obj) : error LNK2001: unresolved external symbol "const tensorflow::data::Metadata::`vftable'" (??_7Metadata@data@tensorflow@@6B@)
2023-04-25T04:32:29.7822169Z bigtable_ops.lo.lib(bigtable_dataset_kernel.obj) : error LNK2001: unresolved external symbol "const tensorflow::data::Metadata::`vftable'" (??_7Metadata@data@tensorflow@@6B@)
2023-04-25T04:32:29.7823223Z avro_atds.lo.lib(atds_dataset_kernels.obj) : error LNK2001: unresolved external symbol "const tensorflow::data::Metadata::`vftable'" (??_7Metadata@data@tensorflow@@6B@)
2023-04-25T04:32:29.7830279Z avro_ops.lo.lib(avro_record_dataset_kernels.obj) : error LNK2001: unresolved external symbol "const tensorflow::data::Metadata::`vftable'" (??_7Metadata@data@tensorflow@@6B@)
2023-04-25T04:32:29.7831211Z kafka_ops.lo.lib(kafka_kernels_deprecated.obj) : error LNK2001: unresolved external symbol "const tensorflow::data::Options::`vftable'" (??_7Options@data@tensorflow@@6B@)
2023-04-25T04:32:29.7832064Z arrow_ops.lo.lib(arrow_dataset_ops.obj) : error LNK2001: unresolved external symbol "const tensorflow::data::Options::`vftable'" (??_7Options@data@tensorflow@@6B@)
2023-04-25T04:32:29.7853097Z bigquery_ops.lo.lib(bigquery_dataset_op.obj) : error LNK2001: unresolved external symbol "const tensorflow::data::Options::`vftable'" (??_7Options@data@tensorflow@@6B@)
2023-04-25T04:32:29.7855005Z bigtable_ops.lo.lib(bigtable_dataset_kernel.obj) : error LNK2001: unresolved external symbol "const tensorflow::data::Options::`vftable'" (??_7Options@data@tensorflow@@6B@)
2023-04-25T04:32:29.7856869Z avro_atds.lo.lib(atds_dataset_kernels.obj) : error LNK2001: unresolved external symbol "const tensorflow::data::Options::`vftable'" (??_7Options@data@tensorflow@@6B@)
2023-04-25T04:32:29.7859178Z avro_ops.lo.lib(avro_record_dataset_kernels.obj) : error LNK2001: unresolved external symbol "const tensorflow::data::Options::`vftable'" (??_7Options@data@tensorflow@@6B@)
2023-04-25T04:32:29.7862665Z avro_atds.lo.lib(atds_dataset_kernels.obj) : error LNK2019: unresolved external symbol "const tensorflow::TensorShapeProto::`vftable'" (??_7TensorShapeProto@tensorflow@@6B@) referenced in function "public: __cdecl tensorflow::TensorShapeProto::TensorShapeProto(void)" (??0TensorShapeProto@tensorflow@@QEAA@XZ)
2023-04-25T04:32:29.7866013Z avro_atds.lo.lib(atds_dataset_kernels.obj) : error LNK2019: unresolved external symbol "const tensorflow::AttrValue::`vftable'" (??_7AttrValue@tensorflow@@6B@) referenced in function "public: __cdecl tensorflow::AttrValue::AttrValue(void)" (??0AttrValue@tensorflow@@QEAA@XZ)
2023-04-25T04:32:29.7868055Z grpc_ops.lo.lib(grpc_kernels.obj) : error LNK2019: unresolved external symbol "const tensorflow::TensorProto::`vftable'" (??_7TensorProto@tensorflow@@6B@) referenced in function "public: __cdecl tensorflow::TensorProto::TensorProto(void)" (??0TensorProto@tensorflow@@QEAA@XZ)
2023-04-25T04:32:29.7871765Z bazel-out\x64_windows-fastbuild\bin\tensorflow_io\python\ops\libtensorflow_io.so : fatal error LNK1120: 5 unresolved externals
```

**Review by kvignesh1420 — APPROVED** [review]
LGTM. The core functionality with initial test cases allows users to leverage this functionality through nightly builds. As per the discussions with @lijuanzhang78, extensive test cases will be added in future PRs. Thank you for your contribution!

**Comment by lijuanzhang78** [comment]
> LGTM. The core functionality with initial test cases allows users to leverage this functionality through nightly builds. As per the discussions with @lijuanzhang78, extensive test cases will be added in future PRs. Thank you for your contribution!

@kvignesh1420 Great! thanks for reviewing this pr. Can we get it merged?

**Comment by oliverhu** [comment]
Thanks everyone!!

File: tensorflow_io/core/kernels/avro/atds/shuffle_handler.h (hunk 3 above)
Comment (plliao) [reply tensorflow_io/core/kernels/avro/atds/shuffle_handler.h:35]: No, it still shuffles in the parsing phase. The problem with Avro is that Avro block does not have metadata to track the offsets of each Avro record in the block. It makes it impossible to jump to a random offset and decode a record at that offset. In other words, we can only read Avro block sequentially. This limitation adds complexity to shuffle. If we want to shuffle Avro records in an Avro block, we have to read all records sequentially and shuffle the intermediate results. It will introduce extra copy which hurts performance. 

In the AvroTensorDataset, we propose a shuffle algorithm that samples the number of records to read from each Avro block and merges the read records from multiple blocks to the outputs. In this way, we can still read Avro records sequentially without extra copy. The Avro block will be kept in memory until they are fully read. Hope it add the context about the shuffle algorithm.

Comment (kvignesh1420) [reply tensorflow_io/core/kernels/avro/atds/shuffle_handler.h:35]: Thanks @plliao
